pkg/util/workqueue/delaying_queue.go

```diff
@@ -0,0 +1,147 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workqueue
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// DelayingInterface is an Interface that can Add an item at a later time.  This makes it easier to
+// requeue items after failures without ending up in a hot-loop.
+type DelayingInterface interface {
+	Interface
+	AddAfter(item interface{}, duration time.Duration)
+}
+
+// NewDelayingQueue constructs a new workqueue with delayed queuing ability
+func NewDelayingQueue() DelayingInterface {
+	return newDelayingQueue(util.RealClock{})
+}
+
+func newDelayingQueue(clock util.Clock) DelayingInterface {
+	ret := &delayingType{
+		Type:          New(),
+		clock:         clock,
+		waitingCond:   sync.NewCond(&sync.Mutex{}),
+		nextCheckTime: clock.Now().Add(60 * time.Second),
+	}
+
+	go ret.waitingLoop()
+	ret.notifyAt(clock.Now().Add(maxWait))
+
+	return ret
+}
+
+type delayingType struct {
+	*Type
+
+	clock util.Clock
+
+	waitingForAdd []waitFor
+	waitingLock   sync.Mutex
+	waitingCond   *sync.Cond
+	nextCheckTime time.Time
+}
+
+type waitFor struct {
+	data    t
+	readyAt time.Time
+}
+
+func (q *delayingType) ShutDown() {
+	q.Type.ShutDown()
+	q.waitingCond.Broadcast()
+}
+
+func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.waitingLock.Lock()
+	defer q.waitingLock.Unlock()
+	waitEntry := waitFor{data: item, readyAt: q.clock.Now().Add(duration)}
+
+	insertionIndex := sort.Search(len(q.waitingForAdd), func(i int) bool {
+		return waitEntry.readyAt.Before(q.waitingForAdd[i].readyAt)
+	})
+
+	tail := q.waitingForAdd[insertionIndex:]
+	q.waitingForAdd = append(q.waitingForAdd[:insertionIndex], waitEntry)
+	q.waitingForAdd = append(q.waitingForAdd, tail...)
+
+	q.notifyAt(waitEntry.readyAt)
+}
+
+const maxWait = 10 * time.Second
+
+func (q *delayingType) waitingLoop() {
+	for {
+		func() {
+			q.waitingCond.L.Lock()
+			defer q.waitingCond.L.Unlock()
+			q.waitingCond.Wait()
+
+			if q.shuttingDown {
+				return
+			}
+
+			func() {
+				q.waitingLock.Lock()
+				defer q.waitingLock.Unlock()
+
+				// always check at least every 10 seconds
+				nextReadyCheck := q.clock.Now().Add(maxWait)
+				nextItemIndex := 0
+
+				for i, queuedItem := range q.waitingForAdd {
+					nextItemIndex = i
+					if queuedItem.readyAt.After(nextReadyCheck) {
+						nextReadyCheck = queuedItem.readyAt
+						break
+					}
+
+					q.Type.Add(queuedItem.data)
+				}
+
+				q.waitingForAdd = q.waitingForAdd[nextItemIndex:]
+				q.notifyAt(nextReadyCheck)
+			}()
+		}()
+	}
+}
+
+func (q *delayingType) notifyAt(nextReadyCheck time.Time) {
+	if nextReadyCheck.After(q.nextCheckTime) {
+		return
+	}
+
+	duration := nextReadyCheck.Sub(q.clock.Now())
+	if duration < 0 || duration > maxWait {
```

Comment:
> what is the rationale behind a max?

Time is weird.  It doesn't move forward on computers, so if you don't have this, something weird will happen someday.  At least that's been my experience in the past.

The max just gives a cadence for a simple sanity check.